### ISSUE TITLE
Join a tournament (S9)

### ISSUE THREAD

**Issue by davidcr01** [davidcr01/WordlePlus#49]
# Description

As a player, is necessary to implement the logic and UI to join a tournament. 

Notice that, when the tournament is closed, is necessary to create the associate 

## Tasks

- [x] Define the controller related to the join of the tournament. (H13-1). Done in #42
- [x] Define the UI of the join of the tournament. (H13-2)

**Comment by davidcr01** [comment]
# Update Report

To manage the game of the tournaments, it is necessary to implement a new method in the Game viewset. This is because in the tournaments the games are created automatically, and not by a player. 
Because of this, the POST and PATCH method executed by the player1 and player2 respectively, implemented in #50 is not useful.

A new patch method has been added:
```
def partial_update_tournament(self, request, *args, **kwargs):
        instance = self.get_object()
        player = getattr(request.user, 'player', None)
        if not player:
            return Response({'error': 'Player not found'}, status=404)

        if player == instance.player1:
            if instance.player1_xp != 0 or instance.player1_time != 0:
                return Response({'error': 'You have already updated your information for this game.'}, status=400)
            opponent = instance.player2
            opponent_xp = instance.player2_xp
            opponent_time = instance.player2_time
        elif player == instance.player2:
            if instance.player2_xp != 0 or instance.player2_time != 0:
                return Response({'error': 'You have already updated your information for this game.'}, status=400)
            opponent = instance.player1
            opponent_xp = instance.player1_xp
            opponent_time = instance.player1_time
        else:
            return Response({'error': 'You do not have permission to update this game.'}, status=403)

        # Data requested change depending on the player who is executing the method
        if instance.player1 == player:
            allowed_fields = ['player1_xp', 'player1_time', 'player1_attempts', 'word']
        elif instance.player2 == player:
            allowed_fields = ['player2_xp', 'player2_time', 'player2_attempts', 'word']
        else:
            return Response({'error': 'You do not have permission to update this game.'}, status=403)

        data = {key: request.data.get(key) for key in allowed_fields}
        if 'winner' in data:
            return Response({'error': 'The "winner" field cannot be modified.'}, status=400)

        if instance.word != '':
            data.pop('word')

        serializer = self.get_serializer(instance, data=data, partial=True)
        serializer.is_valid(raise_exception=True)
        serializer.save()

        player_xp = serializer.validated_data.get('player1_xp' if player == instance.player1 else 'player2_xp', 0)
        player_time = serializer.validated_data.get('player1_time' if player == instance.player1 else 'player2_time', 0)

        # Winner is calculated when all data is available
        if (player_xp != 0 and player_time != 0 and opponent_xp != 0 and opponent_time != 0):
            if player_xp > opponent_xp:
                instance.winner = player
                instance.winner.wins_pvp += 1
                instance.winner.save()
                instance.save()
            elif player_xp < opponent_xp:
                instance.winner = opponent
                opponent.wins_pvp += 1
                instance.winner.save()
                instance.save()
            else:
                if player_time <= opponent_time:
                    instance.winner = player
                    instance.winner.wins_pvp += 1
                    instance.winner.save()
                    instance.save()
                else:
                    instance.winner = opponent
                    opponent.wins_pvp += 1
                    instance.winner.save()
                    instance.save()

        if instance.winner is not None:
            return Response({'winner': instance.winner.user.username})
        else:
            return Response({'message': 'Game updated successfully.'})

```
This method is executed by both players. Every player fills its information, and when all the information is stored, the winner is calculated.
Some error cases are checked:
- Existence of both players.
- Avoid modifying the same data twice.
- Avoid setting the winner.
- Avoid changing the word.

Player2 fills its information:
![image](https://github.com/davidcr01/WordlePlus/assets/72193239/ab96a303-a974-4c01-8429-6b4a992a1d0a)

Player1 fills its information:
![image](https://github.com/davidcr01/WordlePlus/assets/72193239/f443c601-7923-4f62-8810-c9d4bae27d47)

And the winner is set. Notice that the word 'beach' passed in the second patch was ignored.
![image](https://github.com/davidcr01/WordlePlus/assets/72193239/a8553444-486a-47db-8d8b-68b24d64e1df)

**Comment by davidcr01** [comment]
# Update Report - Playing the tournament

As the tournament games are different from the multiplayer games (the tournament games are already created, they are not created by any player), is necessary to implement a new page with a new logic to play the tournament games `game-tournament`.

The HTML code is very similar to `respond-game`:
```
<ion-header>
  <ion-toolbar *ngIf="tournamentInfo">
    <ion-title>Game: {{ tournamentInfo.name }}</ion-title>
  </ion-toolbar>
</ion-header>

<ion-content>
  <ng-container *ngIf="player1 && player2">
  <div class="game-info">
    <div class="username">{{ player1 }}</div>
    <img src="../../../assets/icon/vs.png" alt="VS icon">
    <div class="username">{{ player2 }}</div>
  </div>
  </ng-container>

  <ng-container *ngIf="wordLength && !selectedWord">
    <app-wordle-dashboard [WORDS_LENGTH]="wordLength" [isMultiplayer]="true" (gameFinished)="finishGame($event.time, $event.xp, $event.attempts, $event.selectedWord)"></app-wordle-dashboard>
  </ng-container>
  <ng-container *ngIf="wordLength && selectedWord">
    <app-wordle-dashboard [rightGuessString]="selectedWord" [WORDS_LENGTH]="wordLength" [isMultiplayer]="true" (gameFinished)="finishGame($event.time, $event.xp, $event.attempts, $event.selectedWord)"></app-wordle-dashboard>
  </ng-container>
</ion-content>
```
Notice that the `selectedWord `may be passed if the game already has a word assigned. If not, the word will be selected by the `selectRandomWord` of the Wordle Dashboard component.

In this case, both players use the same API method, but with different bodies:
```
async finishGame(time: number, xp: number, attempts: number, selectedWord: string) {
    let gameData: any = {};

    gameData = {
      player1_time: time,
      player1_xp: xp,
      player1_attempts: attempts,
      word: selectedWord,
    };

    if (this.selfUsername === this.player2) {
      gameData = {
        player2_time: time,
        player2_xp: xp,
        player2_attempts: attempts,
        word: selectedWord,
      };
    }

    (await this.apiService.resolveTournamentGame(this.gameId, gameData)).subscribe(
      (response) => {
        console.log('Game resolved successfully', response);
        if (response.message) {
          setTimeout( () => this.showAlert('Amazing!', 'Who will win?'), 2500);
        }
        else {
          if (response.winner === this.selfUsername) {
            setTimeout( () => this.showAlert('Congratulations!', 'You won! You will be in the next round!'), 2500);
          } else {
            setTimeout( () => this.showAlert('Bad news!', 'You lost. Try next time!'), 2500);
          }
        }
      },
      (error) => {
        console.log('Game could not be resolved', error);
      }
    );
  }
```
Same conditions errors are checked as `create-game` and `respond-game`
